In [7]:
import pandas as pd
from sqlalchemy import create_engine

In [9]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [10]:
pd.io.sql.read_sql('SELECT * FROM books', con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [11]:
pd.io.sql.read_sql('SELECT * FROM authors', con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [12]:
pd.io.sql.read_sql('SELECT * FROM publishers', con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [13]:
pd.io.sql.read_sql('SELECT * FROM ratings', con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [14]:
pd.io.sql.read_sql('SELECT * FROM reviews', con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

## Анализ баз данных

Посмотрим сколько у нас вышло книг после 1 января 2000 года. 

In [15]:
pd.io.sql.read_sql("SELECT COUNT(*) FROM books WHERE publication_date > '2000-01-01'", con = engine)

count
0    819

## Вывод:

В Результате запроса мы установили, что у нас 819 книг вышло после 1 янваля 2000 год

Определим для каждой книги количество обзоров и среднюю оценку 

In [17]:
query = '''
SELECT
    books.title AS title,
    books.book_id AS id,
    AVG(ratings.rating) AS avg_rating,
    COUNT(DISTINCT reviews.review_id) AS cnt_review
FROM
    books
INNER JOIN ratings ON ratings.book_id = books.book_id
INNER JOIN reviews ON reviews.book_id = books.book_id
GROUP BY 
    books.title,
    books.book_id
ORDER BY
    cnt_review DESC
limit 10
'''
pd.io.sql.read_sql(query, con = engine)

title   id  avg_rating  \
0                            Twilight (Twilight  #1)  948    3.662500   
1             The Da Vinci Code (Robert Langdon  #2)  696    3.830508   
2                                    Eat  Pray  Love  207    3.395833   
3                                      The Alchemist  627    3.789474   
4                             The Catcher in the Rye  673    3.825581   
5  The Curious Incident of the Dog in the Night-Time  695    4.081081   
6  Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
7  Harry Potter and the Chamber of Secrets (Harry...  299    4.287500   
8                          Outlander (Outlander  #1)  497    4.125000   
9                                     The Book Thief  656    4.264151   

   cnt_review  
0           7  
1           6  
2           6  
3           6  
4           6  
5           6  
6           6  
7           6  
8           6  
9           6

## Вывод:

В Результате запроса мы вывели список книг c числом обзоров и оценкой. Чтобы таблица была не слишком большая, мы вывели первые 10 результатов.

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры. 

In [18]:
query = '''
SELECT
    publishers.publisher AS publisher_name,
    COUNT(books.book_id) AS cnt_books
FROM
    books
INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
GROUP BY 
    publishers.publisher
HAVING
    SUM(books.num_pages) > 50
ORDER BY
    cnt_books DESC
limit 1
'''
pd.io.sql.read_sql(query, con = engine)

publisher_name  cnt_books
0  Penguin Books         42

## Вывод:
    
В Результате запроса мы определили, что издательство Penguin Books выпускает больше всего книг, толщиной более 50 страниц.

Определим автора с самой высокой средней оценкой книг — при учете только книг с 50 и более оценками

In [19]:
query = '''
SELECT 
    SUBQ.author AS author, 
    AVG(SUBQ.avg_rating) AS avg_rating
FROM 
    (SELECT
        authors.author as author,
        books.book_id AS id,
        AVG(ratings.rating) AS avg_rating,
        COUNT(ratings.rating) AS cnt_rating
FROM 
    authors
INNER JOIN books ON books.author_id = authors.author_id
INNER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY 
    author,
    id
HAVING
    COUNT(ratings.rating) >= 50) AS SUBQ
GROUP BY
    author
ORDER BY
    avg_rating DESC
limit 10
'''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0                         J.K. Rowling/Mary GrandPré    4.283844
1                  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                     J.R.R. Tolkien    4.258446
3                                  Louisa May Alcott    4.192308
4                                       Rick Riordan    4.080645
5                                    William Golding    3.901408
6                                      J.D. Salinger    3.825581
7           Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8  William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                          Dan Brown    3.754540

## Вывод:
    
Автором с самой высокой средней оценкой книг является - J.K. Rowling/Mary GrandPré

Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [20]:
query = '''
SELECT
    AVG(SUBQ.count) AS mean
FROM
    (SELECT COUNT(*)
        FROM reviews
            WHERE username in
                (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating) > 50)
            GROUP BY username)
            AS SUBQ
'''
pd.io.sql.read_sql(query, con = engine)

mean
0  24.333333

## Вывод:
    
Среднее количество обзоров от пользователей, которые поставили больше 50 оценокв - 24.33